src/Data/Text/Utf8/AhoCorasick/Automaton.hs

```diff
@@ -374,67 +379,137 @@ uAt = UVector.unsafeIndex
+        -- input code units.
+        CaseSensitive ->
+          if cu < asciiCount && state == initialState
+            then lookupRootAsciiTransition (offset + 1) (remaining - 1) acc cu
+            else followCodeUnit (offset + 1) (remaining - 1) acc cu state
+        -- Code point is a single byte ==> It's ASCII, no need to call toLower; we can just use maths :)
+        -- We could check < asciiCount here as well, since there are no UTF-8 code unit sequences starting with 10xxxxxx
+        _ | cu < 0xc0 ->
+          let
+            !cu' = if cu >= fromIntegral (Char.ord 'A') && cu <= fromIntegral (Char.ord 'Z') then cu + 0x20 else cu
+          in
+            if state == initialState
+              then lookupRootAsciiTransition (offset + 1) (remaining - 1) acc cu'
+              else followCodeUnit (offset + 1) (remaining - 1) acc cu' state
+        -- Code point is two bytes ==> decode and lowercase
+        _ | cu < 0xe0 -> followLowerCodePoint (offset + 2) (remaining - 2) acc (Utf8.decode2 cu $ indexTextArray u8data $ offset + 1) state
+        -- Code point is three bytes ==> decode and lowercase
+        _ | cu < 0xf0 -> followLowerCodePoint (offset + 3) (remaining - 3) acc (Utf8.decode3 cu (indexTextArray u8data $ offset + 1) (indexTextArray u8data $ offset + 2)) state
+        -- Code point is four bytes ==> definitely outside bmp
+        -- NOTE: There are code points that have 3 code units but are not in the BMP, i.e.
+        -- this code does not have exactly the same behavior as the old version.
+        -- TODO: Now that we are handling code points with multiple code units anyways, we could just toLower here too.
+        _ -> follow4CodeUnits (offset + 4) (remaining - 4) acc cu (indexTextArray u8data $ offset + 1) (indexTextArray u8data $ offset + 2) (indexTextArray u8data $ offset + 3) state
+      where
+        !cu = indexTextArray u8data offset
+
+    -- Lowers the given code point, translates it back into code units and follows those.
+    {-# INLINE followLowerCodePoint #-}
+    followLowerCodePoint :: Int -> Int -> a -> Int -> State -> a
+    followLowerCodePoint !offset !remaining !acc !cp !state
+      | lowerCp < 0x80 = followCodeUnit offset remaining acc (fromIntegral lowerCp) state
+      | lowerCp < 0x800 = follow2CodeUnits offset remaining acc (0xc0 .|. fromIntegral (lowerCp `shiftR` 6)) (0x80 .|. fromIntegral (lowerCp .&. 0x3f)) state
+      | lowerCp < 0x10000 = follow3CodeUnits offset remaining acc (0xe0 .|. fromIntegral (lowerCp `shiftR` 12)) (0x80 .|. fromIntegral ((lowerCp `shiftR` 6) .&. 0x3f)) (0x80 .|. fromIntegral (lowerCp .&. 0x3f)) state
+      | otherwise = error "please handle code points with 4 code units separately"
+      where
+        !lowerCp = Char.ord $ Char.toLower $ Char.chr cp
+
+    -- | Follows 0-4 code units packed into a Word32.
+    {-# INLINE followCodeUnits #-}
+    followCodeUnits :: Int -> Int -> a -> Word32 -> State -> a
```

Comment:
The comment is incorrect (I will update it); this function handles 1-4 code units, the check for zero is performed in `lookupTransition`. Since the null byte is never a part of a sequence of code units of length longer than one, it is handled correctly as well.